source/extensions/filters/http/geoip/geoip_provider_config.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include "envoy/config/typed_config.h"
+#include "envoy/extensions/filters/http/geoip/v3/geoip.pb.h"
+#include "envoy/extensions/filters/http/geoip/v3/geoip.pb.validate.h"
+#include "envoy/network/address.h"
+#include "envoy/protobuf/message_validator.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Geoip {
+
+class Driver {
+public:
+  virtual ~Driver() = default;
+
+  /**
+   *  Looks up a city that is associated with the given IP address.
+   *
+   *  @param   address IP address (IPv4, IPv6) to perform lookup for.
+   *  @returns city for the IP address or absl::nullopt if address is not found in the geolocation
+   * database.
+   */
+  virtual const absl::optional<std::string>&
```

Comment:
Seems unsafe for all these functions to be returning references - if there's any kind of dynamic database reload possible, or if the results for an implementation aren't all held in memory at once, the reference could become invalid before it's used. (References wouldn't work with an RPC-based provider, for example.)

Reply:
good point, missed it